src/tests/simple_pipeline.c

```diff
@@ -0,0 +1,312 @@
+/**
+ * (C) Copyright 2020-2021 Intel Corporation.
+ *
+ * SPDK-License-Identifier: BSD-2-Clause-Patent
+ */
+
+/*
+ * This provides a simple example for how to use the data filter capabilities
+ * in DAOS.
+ */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/types.h>
+#include <daos.h>
+
+/** daos info */
+static daos_handle_t	poh; /** pool */
+static daos_handle_t	coh; /** container */
+static daos_handle_t	oh;  /** object */
+
+#define	ASSERT(cond, ...)					\
+do {								\
+	if (!(cond)) {						\
+		fprintf(stderr, __VA_ARGS__);			\
+		exit(1);					\
+	}							\
+} while (0)
+
+/** DB info */
+#define NR_IODS		3
+#define STRING_LEN	10
+static char	*fields[NR_IODS]	= {"Owner", "Species", "Sex"};
+
+
+void
+insert_example_records(void)
+{
+	int		rc;
+	d_iov_t		dkey;
+	d_sg_list_t	sgls[NR_IODS];
+	d_iov_t		iovs[NR_IODS];
+	daos_iod_t	iods[NR_IODS];
+	uint32_t	i, j;
+
+	char	name[8][STRING_LEN]	= {"Slim\0",
+					   "Buffy\0",
+					   "Claws\0",
+					   "Whistler\0",
+					   "Chirpy\0",
+					   "Browser\0",
+					   "Fang\0",
+					   "Fluffy\0"};
+	char	owner[8][STRING_LEN]	= {"Benny\0",
+					   "Harold\0",
+					   "GWen\0",
+					   "Gwen\0",
+					   "Gwen\0",
+					   "Diane\0",
+					   "Benny\0",
+					   "Harold\0"};
+	char	species[8][STRING_LEN]	= {"snake\0",
+					   "dog\0",
+					   "cat\0",
+					   "bird\0",
+					   "bird\0",
+					   "dog\0",
+					   "dog\0",
+					   "cat\0"};
+	char	sex[8][STRING_LEN]	= {"m\0",
+					   "f\0",
+					   "m\0",
+					   "m\0",
+					   "f\0",
+					   "m\0",
+					   "m\0",
+					   "f\0"};
+	char	**data[NR_IODS];
+
+	data[0] = (char **)owner;
+	data[1] = (char **)species;
+	data[2] = (char **)sex;
+
+	for (i = 0; i < 8; i++) { /** records */
+		/** set dkey for record */
+		d_iov_set(&dkey, name[i], STRING_LEN);
+		for (j = 0; j < NR_IODS; j++) { /** fields */
+			/** akeys */
+			sgls[j].sg_nr		= 1;
+			sgls[j].sg_nr_out	= 0;
+			sgls[j].sg_iovs		= &iovs[0];
+			d_iov_set(&iovs[0], data[j][i], STRING_LEN);
+
+			d_iov_set(&iods[j].iod_name, (void *)fields[j],
+					strlen(fields[j]));
+			iods[j].iod_nr		= 1;
+			iods[j].iod_size	= STRING_LEN;
+			iods[j].iod_recxs	= NULL;
+			iods[j].iod_type	= DAOS_IOD_SINGLE;
+		}
+		rc = daos_obj_update(oh, DAOS_TX_NONE, 0, &dkey, NR_IODS, iods,
+					sgls, NULL);
+		ASSERT(rc == 0, "Obj update failed with %d", rc);
+	}
+}
+
+/**
+ * Build filter "Owner == Benny"
+ */
+void
+build_filter_one(daos_pipeline_t *pipeline)
+{
+	daos_pipeline_filter_t	akey_ft, eqfunc_ft, const_ft;
+	char			constant[STRING_LEN] = "Benny\0";
+	daos_pipeline_node_t	comp_eq_node;
+
+	/** akey for filter */
+	akey_ft.filter_type = DAOS_FILTER_AKEY;
+	akey_ft.data_type   = DAOS_FILTER_TYPE_STRING;
+	akey_ft.num_params  = 0;
+	d_iov_set(&(akey_ft.akey), "Owner", 5);
+	akey_ft.data_offset = 0;
+	akey_ft.data_len    = STRING_LEN;
+
+	/** constant for filter */
```

Comment:
so one thing im not sure about here:
in this use case, the value under the akey "Owner" is always a single value.
what if it's an array value, and you want to check the partial extent value of that array?
for example, in POSIX, we have an akey called "INODE" that has (byte) array values that serializes several things, like for example the mode, permissions, ctime, mtime, etc. 
and let's say i want all dkey entries there with ctime > a certain time stamp. we know ctime is stored at offset 16 in the array and is of size 8 bytes. so what we want is to read the array value under INODE, only from offset 16 and size 8, compare only those 8 bytes in the array value to the passed in timestamp, and return if the condition matches.

from what you have i see we are missing the type if the akey value, like you always assume it's single value? and also specifying a daos_recx_t in case it's an array value to know what part of the array value to read and compare against.

Reply:
> in this use case, the value under the akey "Owner" is always a single value.

doesn't have to be. Here we are only indicating the key for the akey

> for example, in POSIX, we have an akey called "INODE" that has (byte) array values that serializes several things, like for example the mode, permissions, ctime, mtime, etc.
and let's say i want all dkey entries there with ctime > a certain time stamp. we know ctime is stored at offset 16 in the array and is of size 8 bytes. so what we want is to read the array value under INODE, only from offset 16 and size 8, compare only those 8 bytes in the array value to the passed in timestamp, and return if the condition matches.

You can set `data_offset=16` and `data_len=8`. In the case of this simple example, of course, the akey is assumed to be a single whole value, so we pass data_offset=0.

If this does not work for your suggestion, we can always add and extra filter type. Something like `DAOS_FILTER_AKEY_ARRAY` or something like that, to indicate that the data for this akey is accessed as an array, and not as a single value. But we don't need to go there if data_offset+data_len is enough, IMHO.

**NOTE:** reading about `daos_recx_t` I can see that it is pretty much data_offset (interpreted here as indice of first record in the extend) and data_len (interpreted here as number of contiguous records in the extend). I guess, the current design  (data_offset+data_len) will only work if rx_nr  in  here means number of bytes. Otherwise, it won't work. 

What do you think?

Reply:
i see. makes sense.
for single vs array value, one cannot fetch a part of a single value. so you would need to set the value type in the iod anyway when selecting the akey, so probably this is fine and you don't need to do that again. i just wanted to confirm that you can select an extent in case it's an array value.